contracts/LosslessControllerV3.sol

```diff
@@ -381,36 +387,62 @@ contract LosslessControllerV3 is Initializable, ContextUpgradeable, PausableUpgr
     /// @param recipient Address to add the locks
     function _enqueueLockedFunds(ReceiveCheckpoint memory checkpoint, address recipient) private {
         LocksQueue storage queue;
-        queue = tokenScopedLockedFunds[msg.sender].queue[recipient];
-        if (queue.lockedFunds[queue.last].timestamp == checkpoint.timestamp) {
-            queue.lockedFunds[queue.last].amount += checkpoint.amount;
-        } else {
+        queue = tokenScopedLockedFunds[ILERC20(msg.sender)].queue[recipient];
+
+        uint i = queue.first;
+        uint lastItem = queue.last;
+        
+        ReceiveCheckpoint storage lastCheckpoint = queue.lockedFunds[lastItem];
+
+        if (lastCheckpoint.timestamp < checkpoint.timestamp) {
+            // Most common scenario where the item goes at the end of the queue
+            queue.lockedFunds[lastItem + 1] = checkpoint;
             queue.last += 1;
-            queue.lockedFunds[queue.last] = checkpoint;
+        } else if (lastCheckpoint.timestamp == checkpoint.timestamp) {
+            // Second most common scenario where the timestamps are the same so the amount adds up
+            lastCheckpoint.amount += checkpoint.amount;
+        } else {
+            // Edge case: The item has nothing to do with the last item and is not an item that goes at the end of the queue
+            // so we have to iterate in order to rearrange and place the item where it goes.
+            
+            while (i <= lastItem) {
```

Comment:
This operation is extra gas intensive. From what I understand this is the case when tokenLockTimeFrame was changed to a lower value resulting in checkpoint that expires earlier that the last one. Maybe in those cases when we have such a checkpoint, let's just merge it with the last checkpoint so that it expires at the same time as the latest checkpoint. So for example if last checkpoint in the queue expires in 10 minutes, but the new tokenLockTimeFrame is 2 minutes, we would still make this new checkpoints expire in 10 minutes. Once these this last checkpoint is < 2 minutes, checkpoints would once again start queueing in the end of the list.